From 5b46fc6bc0481cdd10db74841114b96ef5c108c1 Mon Sep 17 00:00:00 2001
From: Takanari Hayama <taki@igel.co.jp>
Date: Thu, 1 Mar 2012 17:23:04 +0900
Subject: [PATCH 77/80] rcarfb: register framebuffer after DU has started.

There could be the case where rcarfb is called before DU has started.
If DU is started before initialization, it may lock the bus which
may freeze the whole system. This change avoids such risk.

Signed-off-by: Katsuya Matsubara <matsu@igel.co.jp>
---
 drivers/video/rcarfb.c |   28 ++++++++++++++--------------
 1 files changed, 14 insertions(+), 14 deletions(-)

diff --git a/drivers/video/rcarfb.c b/drivers/video/rcarfb.c
index 2249bb3..5b7562a 100644
--- a/drivers/video/rcarfb.c
+++ b/drivers/video/rcarfb.c
@@ -1098,9 +1098,23 @@ static int __devinit rcar_du_probe(struct platform_device *pdev)
 		 priv->planeconfig[i].y = 0;
 		 priv->planeconfig[i].pitch = info->fix.line_length;
 		 priv->planeconfig[i].phys = info->fix.smem_start + info->var.yoffset * info->fix.line_length + info->var.xoffset * (info->var.bits_per_pixel/8);
+	}
+
+	du_write(priv, DSYSR, (du_read(priv, DSYSR) & ~0x100) | 0x200 );
 
+	for (i=0; i<priv->dispdev->fb_num; i++) {
+		/* plane setting */
+		du_set_plane_mode_planeconfig( i, priv, &priv->planeconfig[i] );
+	}
 
+	error = rcar_du_start(priv);
+	if (error) {
+		dev_err(&pdev->dev, "unable to start hardware\n");
+		goto err1;
+	}
 
+	for (i=0; i<priv->dispdev->fb_num; i++) {
+		info = priv->info[i];
 		error = register_framebuffer(info);
 		if (error < 0)
 			goto err1;
@@ -1112,20 +1126,6 @@ static int __devinit rcar_du_probe(struct platform_device *pdev)
 			 (int) info->var.xres,
 			 (int) info->var.yres,
 			 info->var.bits_per_pixel);
-
-	}
-
-	du_write(priv, DSYSR, (du_read(priv, DSYSR) & ~0x100) | 0x200 );
-
-	for (i=0; i<priv->dispdev->fb_num; i++) {
-		/* plane setting */
-		du_set_plane_mode_planeconfig( i, priv, &priv->planeconfig[i] );
-	}
-
-	error = rcar_du_start(priv);
-	if (error) {
-		dev_err(&pdev->dev, "unable to start hardware\n");
-		goto err1;
 	}
 
 	return 0;
-- 
1.7.0.4

